azure_functions_worker/extension.py

```diff
@@ -0,0 +1,278 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+from types import ModuleType
+from typing import Any, Callable, List, Optional
+import sys
+import importlib
+import logging
+import functools
+from .utils.common import is_python_version
+from .utils.wrappers import enable_feature_by
+from .constants import (
+    PYTHON_ISOLATE_WORKER_DEPENDENCIES,
+    PYTHON_ENABLE_WORKER_EXTENSIONS,
+    PYTHON_ENABLE_WORKER_EXTENSIONS_DEFAULT,
+    PYTHON_ENABLE_WORKER_EXTENSIONS_DEFAULT_39
+)
+from .logging import logger, SYSTEM_LOG_PREFIX
+
+
+# Extension Hooks
+FUNC_EXT_POST_FUNCTION_LOAD = "post_function_load"
+FUNC_EXT_PRE_INVOCATION = "pre_invocation"
+FUNC_EXT_POST_INVOCATION = "post_invocation"
+APP_EXT_POST_FUNCTION_LOAD = "post_function_load_app_level"
+APP_EXT_PRE_INVOCATION = "pre_invocation_app_level"
+APP_EXT_POST_INVOCATION = "post_invocation_app_level"
+
+
+class ExtensionManager:
+    """This marks if the ExtensionManager has already proceeded a detection,
+    if so, the sdk will be cached in ._extension_enabled_sdk
+    """
+    _is_sdk_detected: bool = False
+
+    """This is a cache of azure.functions module that supports extension
+    interfaces. If this is None, that mean the sdk does not support extension.
+    """
+    _extension_enabled_sdk: Optional[ModuleType] = None
+
+    @classmethod
+    @enable_feature_by(
+        flag=PYTHON_ENABLE_WORKER_EXTENSIONS,
+        flag_default=(
+            PYTHON_ENABLE_WORKER_EXTENSIONS_DEFAULT_39 if
+            is_python_version('3.9') else
+            PYTHON_ENABLE_WORKER_EXTENSIONS_DEFAULT
+        )
+    )
+    def function_load_extension(cls, func_name, func_directory):
+        """Helper to execute function load extensions. If one of the extension
+        fails in the extension chain, the rest of them will continue, emitting
+        an error log of an exception trace for failed extension.
+
+        Parameters
+        ----------
+        func_name: str
+            The name of the trigger (e.g. HttpTrigger)
+        func_directory: str
+            The folder path of the trigger
+            (e.g. /home/site/wwwroot/HttpTrigger).
+        """
+        sdk = cls._try_get_sdk_with_extension_enabled()
+        if sdk is None:
+            return
+        else:
+            cls._info_discover_extension_list(func_name, sdk)
+
+        # Invoke function hooks
+        funcs = sdk.ExtensionMeta.get_function_hooks(func_name)
+        cls._safe_execute_function_load_hooks(
+            funcs, FUNC_EXT_POST_FUNCTION_LOAD, func_name, func_directory
+        )
+
+        # Invoke application hook
+        apps = sdk.ExtensionMeta.get_application_hooks()
+        cls._safe_execute_function_load_hooks(
+            apps, APP_EXT_POST_FUNCTION_LOAD, func_name, func_directory
+        )
```

Comment:
I am assuming this is being called for each function load request - does that mean that this piece is going to be run for all the functions and result in the same?

Reply:
Do you check if this has been processed before?

Reply:
Yes, so a single implementation in AppExtension.post_function_load_request_app_level will be applied to all functions when they get loaded.
![image](https://user-images.githubusercontent.com/48038149/113218647-bd7d6b00-9234-11eb-9f84-81af12af348a.png)